src/data_store/table.rs

```diff
@@ -0,0 +1,105 @@
+use std::sync::Arc;
+
+use chrono::NaiveDateTime;
+use datafusion::arrow::array::{BooleanArray, Int64Array, RecordBatch, StringArray};
+use datafusion::prelude::*;
+use datafusion::arrow::datatypes::{DataType, Field, Schema};
+use serde::{Deserialize, Serialize};
+use sqlx::{types::Json, PgPool};
+
+use crate::utils::{LOAD_BALANCER_NAME, TABLE_NAME};
+use super::DataStoreError;
+
+#[derive(Serialize, Deserialize, Debug, sqlx::FromRow)]
+pub struct Table {
+    pub id: i64,
+    pub server_name: String,
+    pub worker_name: String,
+    pub port_name: String,
+    pub active: bool,
+    pub info: Option<Json<Info>>, 
+    pub inserted_at: NaiveDateTime,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Info {
+    pub cnt_conns: i32,
+}
+
+impl Table {
+    pub fn schema() -> Schema {
+        Schema::new(vec![
+            Field::new("id", DataType::Int64, false),
+            Field::new("server_name", DataType::Utf8, true),
+            Field::new("worker_name", DataType::Utf8, true),
+            Field::new("port_name", DataType::Utf8, true),
+            Field::new("active", DataType::Boolean, true),
+            Field::new("info", DataType::Utf8, true),
+            Field::new("inserted_at", DataType::Utf8, true),
+        ])
+    }
+}
+
+impl Table {
+    async fn fetch(pool: PgPool) -> Result<Vec<Self>, DataStoreError> {
+        let sql = format!("select * from {} where server_name = '{}' and active is true", TABLE_NAME, LOAD_BALANCER_NAME);
+        let query = sqlx::query_as::<_, Self>(&sql);
+        let data = query.fetch_all(&pool).await?;
+
+        Ok(data)
+    }
+
+    fn to_df(ctx: SessionContext, records: &mut Vec<Self>) -> Result<DataFrame, DataStoreError> {
```

Comment:
Seems this method does more than just transforming the Table to DataFrames, it also inserts into context.

I mean, it works, just feels it's breaking a couple of SOLID principles, and as your app evolves you will find yourself refactoring this often due to that.

That's my personal opinion of course.

Reply:
I haven't thought about breaking SOLID principles here, so great that you've found this. This method is expected to do only one thing: converting data from postgres to dataframe. That's it. It doesn't insert anything. SessionContext is needed here because it's the main interface to run. But tt's possible just to create SessionContext inside to_df() method and without passing SessionContext as parameter. I'll try to get over From or TryForm traits.